daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -2178,48 +2178,50 @@ private[lf] object SBuiltin {
         }
       case None =>
-        machine.lookupContract(coid) { case V.ContractInstance(srcTemplateId, coinstArg) =>
-          val (upgradingIsEnabled, destTemplateId) = optTargetTemplateId match {
+        machine.lookupContract(coid) { case V.ContractInstance(srcTmplId, coinstArg) =>
+          val (upgradingIsEnabled, dstTmplId) = optTargetTemplateId match {
             case Some(tycon) =>
               (true, tycon)
             case None =>
-              (false, srcTemplateId) // upgrading not enabled; import at source type
+              (false, srcTmplId) // upgrading not enabled; import at source type
           }
-          machine.ensurePackageIsLoaded(
-            destTemplateId.packageId,
-            language.Reference.Template(destTemplateId),
-          ) { () =>
-            importValue(machine, destTemplateId, coinstArg) { templateArg =>
-              getContractInfo(machine, coid, destTemplateId, templateArg, keyOpt) { contract =>
-                ensureContractActive(machine, coid, contract.templateId) {
-
-                  machine.checkContractVisibility(coid, contract)
-                  machine.enforceLimitAddInputContract()
-                  machine.enforceLimitSignatoriesAndObservers(coid, contract)
-
-                  val src: TypeConName = srcTemplateId
-                  val dest: TypeConName = destTemplateId
-
-                  // In Validation mode, we always call validateContractInfo
-                  // In Submission mode, we only call validateContractInfo when src != dest
-                  val needValidationCall: Boolean =
-                    if (machine.validating) {
-                      upgradingIsEnabled
+          if (srcTmplId.qualifiedName != dstTmplId.qualifiedName)
+            Control.Error(
+              IE.WronglyTypedContract(coid, dstTmplId, srcTmplId)
+            )
+          else
+            machine.ensurePackageIsLoaded(
+              dstTmplId.packageId,
+              language.Reference.Template(dstTmplId),
+            ) { () =>
+              importValue(machine, dstTmplId, coinstArg) { templateArg =>
+                getContractInfo(machine, coid, dstTmplId, templateArg, keyOpt) { contract =>
+                  ensureContractActive(machine, coid, contract.templateId) {
+
+                    machine.checkContractVisibility(coid, contract)
+                    machine.enforceLimitAddInputContract()
+                    machine.enforceLimitSignatoriesAndObservers(coid, contract)
+
+                    // In Validation mode, we always call validateContractInfo
+                    // In Submission mode, we only call validateContractInfo when src != dest
+                    val needValidationCall: Boolean =
+                      if (machine.validating) {
+                        upgradingIsEnabled
+                      } else {
+                        upgradingIsEnabled && (srcTmplId.packageId != dstTmplId.packageId)
```

Comment:
I think it would be clearer to just use equality on the identifier (https://github.com/digital-asset/daml/blob/714af3d7ee6f0ee26c9a9d092754b6f32670cac1/daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala#L235).